app/src/main/java/tech/relaycorp/gateway/domain/courier/GenerateCargo.kt

```diff
@@ -1,11 +1,69 @@
 package tech.relaycorp.gateway.domain.courier
 
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.common.toZonedDateTime
+import tech.relaycorp.gateway.data.database.ParcelCollectionDao
+import tech.relaycorp.gateway.data.database.StoredParcelDao
+import tech.relaycorp.gateway.data.disk.DiskMessageOperations
+import tech.relaycorp.gateway.data.disk.MessageDataNotFoundException
+import tech.relaycorp.gateway.data.model.ParcelCollection
+import tech.relaycorp.gateway.data.model.RecipientLocation
+import tech.relaycorp.gateway.data.model.StoredParcel
+import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
+import tech.relaycorp.relaynet.messages.ParcelCollectionAck
+import tech.relaycorp.relaynet.messages.payloads.CargoMessageWithExpiry
+import tech.relaycorp.relaynet.messages.payloads.batch
 import java.io.InputStream
+import java.util.logging.Level
 import javax.inject.Inject
 
 class GenerateCargo
-@Inject constructor() {
+@Inject constructor(
+    private val storedParcelDao: StoredParcelDao,
+    private val parcelCollectionDao: ParcelCollectionDao,
+    private val diskMessageOperations: DiskMessageOperations
+) {
 
-    // TODO: implementation
-    suspend fun generate(): Iterable<InputStream> = emptyList()
+    suspend fun generate(): Iterable<() -> InputStream> =
```

Comment:
Note that this isn't generating cargoes: It's generating the payload for each cargo, but it's not encrypting the payload nor is it creating the actual cargo itself.

This function should be doing something like:

```
suspend fun generateCargoes(): Sequence<ByteArray> = (generatePCAs() + retrieveParcels())
    .batch()
    .map {
        val cargoMessageSetSerialized = it.cargoMessageSet.encrypt(publicGatewayCert)
        val creationDate = ZonedDateTime.now()
        val ttl = ChronoUnit.SECONDS.between(creationDate, it.latestMessageExpiryDate)
        val cargo = Cargo(
            "https://public-gateway.com",
            privateGatewayCert,
            cargoMessageSetSerialized,
            creationDate = creationDate,
            ttl = ttl
        )
        return cargo.serialize(privateGatewayKeyPair.privateKey)
    }
```

Reply:
Just to make sure, the `messageId` goes null? Or do we generate one?

Reply:
It can be unspecified, in which case the core lib will generate a UUID4.

We might want to generate custom message ids if we want to override a cargo message we've already sent though, but we can discuss that later.